kubernetes/controllers/elasticjob_controller.go

```diff
@@ -0,0 +1,207 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	commonv1 "github.com/kubeflow/common/pkg/apis/common/v1"
+	"github.com/kubeflow/common/pkg/controller.v1/common"
+	logger "github.com/kubeflow/common/pkg/util"
+	"github.com/pytorch/elastic/kubernetes/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/util/workqueue"
+	k8scontroller "k8s.io/kubernetes/pkg/controller"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	controllerName      = "elastic-job-controller"
+	elasticJobRoleLabel = "elastic-job-role"
+)
+
+var (
+	jobOwnerKey           = ".metadata.controller"
+	defaultTTLseconds     = int32(100)
+	defaultCleanPodPolicy = commonv1.CleanPodPolicyNone
+)
+
+// ElasticJobReconciler reconciles a ElasticJob object
+type ElasticJobReconciler struct {
+	jobController common.JobController
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a ElasticJob object and makes changes based on the state read
+// and what is in the ElasticJob.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=elastic.pytorch.org,resources=elasticjobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=elastic.pytorch.org,resources=elasticjobs/status,verbs=get;update;patch
+
+func (r *ElasticJobReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	// Fetch the ElasticJob instance
+	elasticJob := &v1alpha1.ElasticJob{}
+	err := r.Get(context.TODO(), req.NamespacedName, elasticJob)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	log := logger.LoggerForJob(elasticJob)
+	needSync := r.satisfiedExpections(elasticJob)
+	if !needSync {
+		log.Info("reconcile skipped, job does not need to sync")
+		return ctrl.Result{}, nil
```

Comment:
Are these logs available when one does `kubectl get logs`?

Reply:
Yes. These logs will be available if you check controller.